src/components/App.tsx

```diff
@@ -2947,6 +2956,30 @@ class App extends React.Component<ExcalidrawProps, AppState> {
     }
   };
 
+  private onKeyDownFromPointerDownHandler(
+    pointerDownState: PointerDownState,
+  ): (event: KeyboardEvent) => void {
+    return withBatchedUpdates((event: KeyboardEvent) => {
+      if (this.maybeHandleResize(pointerDownState, event)) {
+        return;
+      }
+      this.maybeDragNewGenericElement(pointerDownState, event);
+    });
+  }
+
+  private onKeyUpFromPointerDownHandler(
+    pointerDownState: PointerDownState,
+  ): (event: KeyboardEvent) => void {
+    return withBatchedUpdates((event: KeyboardEvent) => {
+      // Prevents focus from escaping excalidraw tab
+      event.key === KEYS.ALT && event.preventDefault();
+      if (this.maybeHandleResize(pointerDownState, event)) {
+        return;
+      }
+      this.maybeDragNewGenericElement(pointerDownState, event);
+    });
+  }
+
```

Comment:
Do I understand it correctly that these handlers are implemented solely to update the shape right after keypress? As such, when user holds `alt`/`shift`/`ctrl` down during resize/creating these handlers will be called continuously and hence will duplicate the calls unnecessarily? Should we make the `pointerdown` into a `once` listener and rebuind on keyup? I'm not sure if this perf optim is needed, but I'd lean towards "yes" because resize/drag loop is quite a hot path.

Reply:
I tested dragging ~400 elements while holding down shift/ctrl and at times it completely grinds to a halt. Production isn't much better, but it tells us that this path must really be optimized where we can.

https://excalidraw.com/#json=5906421031895040,m27aPbL-d54AQA65WSurVQ

Reply:
> Do I understand it correctly that these handlers are implemented solely to update the shape right after keypress?
Yes
>  As such, when user holds alt/shift/ctrl down during resize/move/creation these handlers will be called continuously and hence will duplicate the calls unnecessarily?
From my experiments, no. The handlers will be called once for every keyDown or keyUp. When a user holds a key and drags/moves/resize an element, `onKeyDownFromPointerDownHandler` and `onKeyUpFromPointerDownHandler` won't be called for every cursor move. So I don't think we need a performance optimization :)

Reply:
I'll check it out and see what I can do.

Reply:
It doesn't fire on mousemove, but rather continuously while you hold the key. Although that code path probably isn't what's causing the bottleneck:

![image](https://user-images.githubusercontent.com/5153846/101290046-53ae0100-3800-11eb-92ae-b87f451050fc.png)

Browser also seems to fire the keydown only in ~30ms intervals so if we keep that handle lean, I guess we're ok.

Reply:
Yeah, I run some more tests, and I also don't think that's causing the bottleneck.

Ok. Let's leave it then.